Unable to build for C++17

Hello,

I'm looking to build a static lib to link into my C++17 project, but I'm running into errors when trying to build the project:

```
$ ./configure.py --cc=gcc --cxxflags="-std=c++17"
...
$ make                                           
g++ -fPIC -fvisibility=hidden -fstack-protector -pthread -std=c++20 -D_REENTRANT  -std=c++17 -Wall -Wextra -Wpedantic -Wstrict-aliasing -Wcast-align -Wmissing-declarations -Wpointer-arith -Wcast-qual -Wzero-as-null-pointer-constant -Wnon-virtual-dtor -Wold-style-cast -Wsuggest-override -Wshadow -Wextra-semi -DBOTAN_IS_BEING_BUILT  -I build/include/public -I build/include/internal  -isystem build/include/external -c src/lib/asn1/alg_id.cpp -o build/obj/lib/asn1_alg_id.o
In file included from src/lib/asn1/alg_id.cpp:8:
In file included from build/include/public/botan/asn1_obj.h:11:
build/include/public/botan/secmem.h:29:4: error: unknown type name 'requires'
   29 |    requires std::is_integral<T>::value || std::is_enum<T>::value
      |    ^
build/include/public/botan/secmem.h:29:34: error: cannot define or redeclare 'value' here because namespace 'Botan' does not enclose namespace 'is_integral'
   29 |    requires std::is_integral<T>::value || std::is_enum<T>::value
      |             ~~~~~~~~~~~~~~~~~~~~~^
build/include/public/botan/secmem.h:29:39: error: expected ';' at end of declaration
   29 |    requires std::is_integral<T>::value || std::is_enum<T>::value
      |                                       ^
      |                                       ;
build/include/public/botan/secmem.h:29:40: error: expected unqualified-id
   29 |    requires std::is_integral<T>::value || std::is_enum<T>::value
      |                                        ^
build/include/public/botan/secmem.h:51:30: error: no template named 'secure_allocator'
   51 | inline bool operator==(const secure_allocator<T>&, const secure_allocator<U>&) {
      |                              ^
build/include/public/botan/secmem.h:51:58: error: no template named 'secure_allocator'
   51 | inline bool operator==(const secure_allocator<T>&, const secure_allocator<U>&) {
      |                                                          ^
build/include/public/botan/secmem.h:56:30: error: no template named 'secure_allocator'
   56 | inline bool operator!=(const secure_allocator<T>&, const secure_allocator<U>&) {
      |                              ^
build/include/public/botan/secmem.h:56:58: error: no template named 'secure_allocator'
   56 | inline bool operator!=(const secure_allocator<T>&, const secure_allocator<U>&) {
      |                                                          ^
build/include/public/botan/secmem.h:61:38: error: use of undeclared identifier 'secure_allocator'
   61 | using secure_vector = std::vector<T, secure_allocator<T>>;
      |                                      ^
build/include/public/botan/secmem.h:63:36: error: use of undeclared identifier 'secure_allocator'
   63 | using secure_deque = std::deque<T, secure_allocator<T>>;
      |                                    ^
build/include/public/botan/secmem.h:67:22: error: no template named 'secure_vector'
   67 | using SecureVector = secure_vector<T>;
      |                      ^
build/include/public/botan/secmem.h:70:1: error: no template named 'secure_vector'
   70 | secure_vector<T> lock(const std::vector<T>& in) {
      | ^
build/include/public/botan/secmem.h:71:11: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++20 extension [-Wc++20-extensions]
   71 |    return secure_vector<T>(in.begin(), in.end());
      |           ^
build/include/public/botan/secmem.h:75:29: error: no template named 'secure_vector'
   75 | std::vector<T> unlock(const secure_vector<T>& in) {
      |                             ^
In file included from src/lib/asn1/alg_id.cpp:8:
build/include/public/botan/asn1_obj.h:155:12: error: no template named 'span' in namespace 'std'
  155 |       std::span<const uint8_t> data() const { return std::span{m_value}; }
      |       ~~~~~^
build/include/public/botan/asn1_obj.h:166:7: error: no template named 'secure_vector'
  166 |       secure_vector<uint8_t> m_value;
      |       ^
build/include/public/botan/asn1_obj.h:155:59: error: no member named 'span' in namespace 'std'
  155 |       std::span<const uint8_t> data() const { return std::span{m_value}; }
      |                                                      ~~~~~^
build/include/public/botan/asn1_obj.h:155:63: error: expected ';' after return statement
  155 |       std::span<const uint8_t> data() const { return std::span{m_value}; }
      |                                                               ^
      |                                                               ;
In file included from src/lib/asn1/alg_id.cpp:10:
In file included from build/include/public/botan/ber_dec.h:12:
build/include/public/botan/data_src.h:124:57: error: expected ')'
  124 |       explicit DataSource_Memory(secure_vector<uint8_t> in) : m_source(std::move(in)), m_offset(0) {}
      |                                                         ^
build/include/public/botan/data_src.h:124:33: note: to match this '('
  124 |       explicit DataSource_Memory(secure_vector<uint8_t> in) : m_source(std::move(in)), m_offset(0) {}
      |                                 ^
build/include/public/botan/data_src.h:124:82: error: use of undeclared identifier 'in'
  124 |       explicit DataSource_Memory(secure_vector<uint8_t> in) : m_source(std::move(in)), m_offset(0) {}
      |                                                                                  ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
1 warning and 20 errors generated.
make: *** [build/obj/lib/asn1_alg_id.o] Error 1
```

Is there anything I'm missing or doing wrong?

Botan version: 3.6.1
OS: M1 macOS

--- a/doc/building.rst
+++ b/doc/building.rst
@@ -17,6 +17,9 @@ Currently systems such as VMS, OS/390, and OS/400 are not supported by the build
 system, primarily due to lack of access and interest.  Please contact the
 maintainer if you would like to build Botan on such a system.
 
+Botan is a C++20 code base, make sure to use an appropriate compiler and
+settings. See also :ref:`support_info`.
+
 Botan's build is controlled by configure.py, which is a `Python
 <https://www.python.org>`_ script. Python 3.x or later is required.
 

--- a/doc/support.rst
+++ b/doc/support.rst
@@ -1,3 +1,5 @@
+.. _support_info:
+
 Support Information
 =======================
 

--- a/src/build-data/buildh.in
+++ b/src/build-data/buildh.in
@@ -285,4 +285,21 @@
     #error "Trying to compile Botan configured as x86_32 with non-x86_32 compiler."
 #endif
 
+/**
+* MSVC does define __cplusplus but pins it at 199711L, because "legacy".
+* Note: There is a compiler switch to enable standard behavior (/Zc:__cplusplus),
+*       but we can't control that in downstream applications.
+*
+* See: https://learn.microsoft.com/en-us/cpp/build/reference/zc-cplusplus
+*/
+#if defined(_MSVC_LANG)
+   #define BOTAN_CPLUSPLUS _MSVC_LANG
+#else
+   #define BOTAN_CPLUSPLUS __cplusplus
+#endif
+
+#if BOTAN_CPLUSPLUS < 202002L
+   #error "Botan 3.x requires at least C++20"
+#endif
+
 #endif

